File: vl2ggplot/R/literals.R
# Render plain JSON-compatible R values (as produced by jsonlite::fromJSON
# with simplifyVector = FALSE) into R source text.
#
# Unlike JS/Python, R distinguishes atomic vectors (c(1, 2, 3)) from lists
# (list(1, "a", list(...))) -- a JSON array becomes a c(...) call only when
# every element is a scalar of the *same* atomic type; otherwise it becomes
# list(...), same as a JSON object always does (as a named list).

.max_line <- 88L

.r_reserved <- c(
  "if", "else", "repeat", "while", "function", "for", "next", "break",
  "TRUE", "FALSE", "NULL", "Inf", "NaN", "NA", "NA_integer_", "NA_real_",
  "NA_character_", "in"
)

is_valid_r_name <- function(name) {
  grepl("^[a-zA-Z.][a-zA-Z0-9._]*$", name) && !(name %in% .r_reserved)
}

render_name <- function(name) {
  if (is_valid_r_name(name)) name else paste0("`", gsub("`", "\\`", name, fixed = TRUE), "`")
}

render_string <- function(x) {
  escaped <- gsub("\\\\", "\\\\\\\\", x)
  escaped <- gsub("\"", "\\\\\"", escaped)
  escaped <- gsub("\n", "\\\\n", escaped)
  paste0("\"", escaped, "\"")
}

render_scalar <- function(x) {
  if (is.null(x)) return("NULL")
  # A non-scalar value here means a JSON array/object cell (e.g. a "flatten"
  # transform's un-flattened array-valued field) reached what's normally a
  # scalar-only path; delegate to the general renderer instead of crashing
  # inside is.na(), which requires length-1 input.
  if (!is_scalar_value(x)) return(format_value(x))
  if (is.na(x)) return("NA")
  if (is.logical(x)) return(if (x) "TRUE" else "FALSE")
  if (is.numeric(x)) return(format(x, scientific = FALSE, trim = TRUE))
  if (is.character(x)) return(render_string(x))
  render_string(as.character(x))
}

is_scalar_value <- function(x) {
  is.null(x) || (length(x) == 1 && (is.logical(x) || is.numeric(x) || is.character(x)) && !is.list(x))
}

is_named_list <- function(x) {
  is.list(x) && !is.null(names(x)) && all(nzchar(names(x)))
}

# All elements are non-list, non-NULL scalars of the same atomic mode.
homogeneous_atomic <- function(items) {
  if (length(items) == 0) return(FALSE)
  if (any(vapply(items, function(v) is.list(v) || is.null(v), logical(1)))) return(FALSE)
  modes <- vapply(items, function(v) if (is.logical(v)) "logical" else if (is.numeric(v)) "numeric" else "character", character(1))
  length(unique(modes)) == 1
}

render_inline <- function(x) {
  if (is_scalar_value(x)) return(render_scalar(x))

  if (is_named_list(x)) {
    parts <- vapply(names(x), function(n) {
      inner <- render_inline(x[[n]])
      if (is.na(inner)) return(NA_character_)
      paste0(render_name(n), " = ", inner)
    }, character(1))
    if (any(is.na(parts))) return(NA_character_)
    return(paste0("list(", paste(parts, collapse = ", "), ")"))
  }

  if (is.list(x)) {
    if (length(x) == 0) return("list()")
    if (homogeneous_atomic(x)) {
      parts <- vapply(x, render_inline, character(1))
      return(paste0("c(", paste(parts, collapse = ", "), ")"))
    }
    parts <- vapply(x, function(v) {
      inner <- render_inline(v)
      if (is.na(inner)) NA_character_ else inner
    }, character(1))
    if (any(is.na(parts))) return(NA_character_)
    return(paste0("list(", paste(parts, collapse = ", "), ")"))
  }

  NA_character_
}

render_multiline <- function(x, indent) {
  pad <- strrep("  ", indent + 1)
  closing_pad <- strrep("  ", indent)

  if (is_named_list(x)) {
    lines <- vapply(names(x), function(n) {
      paste0(pad, render_name(n), " = ", format_value(x[[n]], indent + 1), ",")
    }, character(1))
    lines[length(lines)] <- sub(",$", "", lines[length(lines)])
    return(paste0("list(\n", paste(lines, collapse = "\n"), "\n", closing_pad, ")"))
  }

  if (is.list(x)) {
    if (length(x) == 0) return("list()")
    ctor <- if (homogeneous_atomic(x)) "c" else "list"
    lines <- vapply(x, function(v) paste0(pad, format_value(v, indent + 1), ","), character(1))
    lines[length(lines)] <- sub(",$", "", lines[length(lines)])
    return(paste0(ctor, "(\n", paste(lines, collapse = "\n"), "\n", closing_pad, ")"))
  }

  render_scalar(x)
}

# Render `x` (a value from jsonlite::fromJSON(..., simplifyVector = FALSE))
# as R source, pretty-printing across multiple lines when the single-line
# form would be too long.
format_value <- function(x, indent = 0) {
  inline <- render_inline(x)
  if (!is.na(inline) && nchar(inline) <= (.max_line - indent * 2) && !grepl("\n", inline, fixed = TRUE)) {
    return(inline)
  }
  render_multiline(x, indent)
}

File: vl2altair/literals.py
"""Render plain Python values (as produced by ``json.load``) into Python source text."""

from __future__ import annotations

import keyword

MAX_LINE = 88


def is_safe_identifier(key: object) -> bool:
    """Whether ``key`` can be used as a literal ``name=value`` keyword argument."""
    return (
        isinstance(key, str)
        and key.isidentifier()
        and not keyword.iskeyword(key)
    )


def format_value(value: object, indent: int = 0) -> str:
    """Render a JSON-compatible Python value as Python source code.

    Dicts/lists are pretty-printed across multiple lines when the single-line
    form would be too long; short literals stay inline.
    """
    inline = _format_inline(value)
    if inline is not None and (len(inline) <= MAX_LINE - indent * 4 and "\n" not in inline):
        return inline
    return _format_multiline(value, indent)


def _format_inline(value: object) -> str | None:
    if isinstance(value, dict):
        if not value:
            return "{}"
        parts = []
        for k, v in value.items():
            iv = _format_inline(v)
            if iv is None:
                return None
            parts.append(f"{_format_inline(k)}: {iv}")
        return "{" + ", ".join(parts) + "}"
    if isinstance(value, (list, tuple)):
        if not value:
            return "[]"
        parts = []
        for v in value:
            iv = _format_inline(v)
            if iv is None:
                return None
            parts.append(iv)
        return "[" + ", ".join(parts) + "]"
    return _format_scalar(value)


def _format_scalar(value: object) -> str:
    if value is None:
        return "None"
    if isinstance(value, bool):
        return "True" if value else "False"
    if isinstance(value, (int, float)):
        return repr(value)
    if isinstance(value, str):
        return repr(value)
    # Fallback for anything unexpected (should not normally happen for JSON data).
    return repr(value)


def _format_multiline(value: object, indent: int) -> str:
    pad = "    " * (indent + 1)
    closing_pad = "    " * indent
    if isinstance(value, dict):
        if not value:
            return "{}"
        lines = []
        for k, v in value.items():
            rendered = format_value(v, indent + 1)
            lines.append(f"{pad}{_format_inline(k)}: {rendered},")
        return "{\n" + "\n".join(lines) + f"\n{closing_pad}}}"
    if isinstance(value, (list, tuple)):
        if not value:
            return "[]"
        lines = []
        for v in value:
            rendered = format_value(v, indent + 1)
            lines.append(f"{pad}{rendered},")
        return "[\n" + "\n".join(lines) + f"\n{closing_pad}]"
    return _format_scalar(value)


def try_black_format(source: str) -> str:
    """Best-effort formatting with ``black`` if it is installed; otherwise a no-op."""
    try:
        import black
    except ImportError:
        return source
    try:
        return black.format_str(source, mode=black.Mode())
    except Exception:
        return source
